File: retrocast-client/Retrocast/Sources/Models/Message.swift
import Foundation

struct Message: Codable, Identifiable, Hashable, Sendable {
    let id: Snowflake
    let channelID: Snowflake
    let authorID: Snowflake
    var content: String
    let createdAt: Date
    var editedAt: Date?

    // Joined author fields from server
    let authorUsername: String?
    let authorDisplayName: String?
    let authorAvatarHash: String?

    enum CodingKeys: String, CodingKey {
        case id
        case channelID = "channel_id"
        case authorID = "author_id"
        case content
        case createdAt = "created_at"
        case editedAt = "edited_at"
        case authorUsername = "author_username"
        case authorDisplayName = "author_display_name"
        case authorAvatarHash = "author_avatar_hash"
    }

    /// Display name: prefer display_name, fall back to username.
    var displayName: String {
        authorDisplayName ?? authorUsername ?? "Unknown"
    }
}

File: retrocast/internal/models/message.go
package models

import "time"

type Message struct {
	ID        int64      `json:"id,string"`
	ChannelID int64      `json:"channel_id,string"`
	AuthorID  int64      `json:"author_id,string"`
	Content   string     `json:"content"`
	CreatedAt time.Time  `json:"created_at"`
	EditedAt  *time.Time `json:"edited_at,omitempty"`
}

type MessageWithAuthor struct {
	Message
	AuthorUsername    string  `json:"author_username"`
	AuthorDisplayName string  `json:"author_display_name"`
	AuthorAvatarHash  *string `json:"author_avatar_hash,omitempty"`
}
